File: mlvm/lib/Array/Array.cpp
#include <sstream>

#include "mlvm/lib/Array/Array.h"

namespace mlvm {
namespace array {

const Data &Array::data() const { return data_; }

const Shape &Array::shape() const { return shape_; }

std::string Array::DebugString() const {
  std::stringstream ss;
  ss << *this;
  return ss.str();
}

std::ostream &operator<<(std::ostream &out, const Array &arr) {
  out << "\"" << arr.name_ << "\" ";
  out << arr.shape_;
  return out;
}

}  // namespace array
}  // namespace mlvm

File: mlvm/array/array.go
package array

type Dimension int

type Float float32

type Shape struct {
	Dims []Dimension
}

type Data struct {
	Value []Float
}

type Array struct {
	name  string // Name of the Array
	shape *Shape // The real shape.
	data  *Data  // The backing data.
}

func (arr *Array) Name() string {
	return arr.name
}

func (arr *Array) Shape() *Shape {
	return arr.shape
}

func (arr *Array) Data() *Data {
	return arr.data
}

// func IsScalar(shape *Shape) bool {
// 	return len(shape) == 1 && shape[0] == 1
// }
//
// func ElementCount(shape []Dimension) int {
// 	count := 1
// 	for _, dim := range shape {
// 		count *= int(dim)
// 	}
// 	return count
// }
